File: docs/src/test/scala/docs/home/journal/EventJsonColumn.scala
/*
 * Copyright (C) 2022 - 2025 Lightbend Inc. <https://akka.io>
 */

package docs.home.journal

import akka.persistence.r2dbc.journal.scaladsl.AdditionalColumn
// #additional-column-json
import io.r2dbc.postgresql.codec.Json

class EventJsonColumn extends AdditionalColumn[BlogPost.Event, Json] {

  override val columnName: String = "event_json"

  override def bind(insert: AdditionalColumn.Insert[BlogPost.Event]): AdditionalColumn.Binding[Json] =
    insert.value match {
      case BlogPost.PostAdded(_, title, _) =>
        // a json library would be used here
        val jsonString = s"""{"title": "$title", "published": false}"""
        AdditionalColumn.BindValue(Json.of(jsonString))
      case _: BlogPost.Published =>
        val jsonString = """{"published": true}"""
        AdditionalColumn.BindValue(Json.of(jsonString))
      case _ =>
        AdditionalColumn.BindNull
    }
}
// #additional-column-json

File: docs/src/test/java/jdocs/home/journal/EventJsonColumn.java
/*
 * Copyright (C) 2022 - 2025 Lightbend Inc. <https://akka.io>
 */

package jdocs.home.journal;

import akka.persistence.r2dbc.journal.javadsl.AdditionalColumn;
// #additional-column-json
import io.r2dbc.postgresql.codec.Json;

public class EventJsonColumn extends AdditionalColumn<BlogPost.Event, Json> {
  @Override
  public Class<Json> fieldClass() {
    return Json.class;
  }

  @Override
  public String columnName() {
    return "event_json";
  }

  @Override
  public Binding<Json> bind(Insert<BlogPost.Event> insert) {
    BlogPost.Event event = insert.value();
    if (event instanceof BlogPost.PostAdded) {
      BlogPost.PostAdded added = (BlogPost.PostAdded) event;
      // a json library would be used here
      String jsonString = "{\"title\": \"" + added.title + "\", \"published\": false}";
      return AdditionalColumn.bindValue(Json.of(jsonString));
    } else if (event instanceof BlogPost.Published) {
      String jsonString = "{\"published\": true}";
      return AdditionalColumn.bindValue(Json.of(jsonString));
    } else {
      return AdditionalColumn.bindNull();
    }
  }
}
// #additional-column-json
